contracts/staking/contracts/src/stake/MixinStakeStorage.sol

```diff
@@ -0,0 +1,199 @@
+        internal
+    {
+        // Remove stake from balance
+        IStructs.DelayedBalance memory balance = _syncBalanceDestructive(balancePtr);
+        balance.next = LibSafeDowncast.downcastToUint96(uint256(balance.next).safeAdd(amount));
+        balance.current = LibSafeDowncast.downcastToUint96(uint256(balance.current).safeAdd(amount));
+
+        // update state
+        _storeBalance(balancePtr, balance);
+    }
+
+    /// @dev Burns existing value in a balance.
+    ///      This causes both the `current` and `next` fields to decrease immediately.
+    /// @param balancePtr storage pointer to balance.
+    /// @param amount to mint.
+    function _burnBalance(IStructs.DelayedBalance storage balancePtr, uint256 amount)
+        internal
+    {
+        // Remove stake from balance
+        IStructs.DelayedBalance memory balance = _syncBalanceDestructive(balancePtr);
+        balance.next = LibSafeDowncast.downcastToUint96(uint256(balance.next).safeSub(amount));
+        balance.current = LibSafeDowncast.downcastToUint96(uint256(balance.current).safeSub(amount));
+
+        // update state
+        _storeBalance(balancePtr, balance);
+    }
+
+    /// @dev Increments a balance.
+    ///      Ths updates the `next` field but not the `current` field.
+    /// @param balancePtr storage pointer to balance.
+    /// @param amount to increment by.
+    function _incrementBalance(IStructs.DelayedBalance storage balancePtr, uint256 amount)
+        internal
+    {
+        // Add stake to balance
+        IStructs.DelayedBalance memory balance = _syncBalanceDestructive(balancePtr);
+        balance.next = LibSafeDowncast.downcastToUint96(uint256(balance.next).safeAdd(amount));
+
+        // update state
+        _storeBalance(balancePtr, balance);
+    }
+
+    /// @dev Decrements a balance.
+    ///      Ths updates the `next` field but not the `current` field.
+    /// @param balancePtr storage pointer to balance.
+    /// @param amount to decrement by.
+    function _decrementBalance(IStructs.DelayedBalance storage balancePtr, uint256 amount)
+        internal
+    {
+        // Remove stake from balance
+        IStructs.DelayedBalance memory balance = _syncBalanceDestructive(balancePtr);
+        balance.next = LibSafeDowncast.downcastToUint96(uint256(balance.next).safeSub(amount));
+
+        // update state
+        _storeBalance(balancePtr, balance);
+    }
+
+    /// @dev Stores a balance in storage.
+    /// @param balancePtr points to where `balance` will be stored.
+    /// @param balance to save to storage.
+    function _storeBalance(
+        IStructs.DelayedBalance storage balancePtr,
+        IStructs.DelayedBalance memory balance
+    )
+        private
+    {
+        // note - this compresses into a single `sstore` when optimizations are enabled,
+        // since the StakeBalance struct occupies a single word of storage.
+        balancePtr.lastStored = balance.lastStored;
```

Comment:
Yeah unfortunately that assignment won't work in Solidity. Because `balancePtr` is a storage pointer, the `=` operator changes what it points to but not the underlying value.